docs/rsdl/rapid-pro-rsdl-intro.md

```diff
@@ -3,80 +3,238 @@ id: rsdl-intro
+~~~rsdl
+type fullName
+{
+    firstName: string
+    lastName: string
+}
+~~~
+
+and then use that type in our employee:
 
-## Defining top level single objects
+~~~rsdl
+type employee
+{
+    @key id: integer
+    name: fullName
+    title: string
+}
+~~~
 
-## Querying a graph of objects
+### Defining methods
+RAPID supports functions and actions.
 
-### Types with keys vs types without
+A function takes zero or more input parameters, and returns a value.  Functions must not have side-affects.
 
-### Navigation from one object to another
+We can define a "topEmployees" function on our company:
 
-## Changing data
+~~~rsdl
+type company
+{
+    @key stockSymbol: string
+    name: string
+    incorporated: dateTime
+    employees: [employee]
+    topEmployees(num: integer) : [employee]
+}
+~~~
 
-### create
+topEmployees takes a single integer parameter "num" and returns a collection of employees.
 
-### update
+Functions are invoked using a GET request. Function parameters are passed in the URL.
 
-### delete
+| Request                                                                                              | Comment                                                          |
+| :--------------------------------------------------------------------------------------------------- | :--------------------------------------------------------------- |
+| GET [http://rapid-pro.org/company/topEmployees?num=10](https://jetsons.azurewebsites.net/company/Jetsons.Models.topEmployees(num=10))  | get the company's top 10 employees                             |
+|
+An action takes zero or more input parameters and may or may not return a value.  Actions may have side-affects.
+
+We can define a "youreFired" action on our company that takes a string parameter "reason":
+
+~~~rsdl
+type company
+{
+    @key stockSymbol: string
+    name: string
+    incorporated: dateTime
+    employees: [employee]
+    topEmployees(num: integer): [employee]
+    @action yourFired(reason: string)
+}
+~~~
+
+yourFired has the @action attribute to show that it is an action and may have side-affects. It does not return a value.
+
+Because actions may have side-affects, they are invoked using POST. Their parameters are passed in the body of the request.
+
+| Request                                                                                              | Comment                                                          |
+| :--------------------------------------------------------------------------------------------------- | :--------------------------------------------------------------- |
+| POST http://rapid-pro.org/company/employees/1/yourFired <br/> { "reason": "Embezzlement" }  | invoke the yourFired action on employee with id = 1       |
+|
+
+Actions and functions may also be defined on the service.
+
+~~~rsdl
+jetsons {
+    company: company
+    competitors: [company]
+    currentStockPrice(stockSymbol: string): decimal
```

Comment:
This actually would better be a bound function. 

These examples will be seen as instructions on how to correctly model RAPID services.

Reply:
Yes; I realized this would be better as a bound function, but it was getting late and I didn't have a better example. Let me see what else I can come up with.